### PR TITLE
Implement sql.Scanner and driver.Value interfaces

### PR DESCRIPTION
**Pull request by bruth**
This adds methods to work seamlessly with SQL statements.

/cc @peterbourgon @tsenart

### CODE REVIEW

**Comment by bruth** [comment]
CI build failed due to a Coveralls error:

```
$ $HOME/gopath/bin/goveralls -coverprofile=cov.out -service=travis-ci -repotoken $COVERALLS_TOKEN
Bad response status from coveralls: 422
{"message":"Couldn't find a repository matching this job.","error":true}
The command "$HOME/gopath/bin/goveralls -coverprofile=cov.out -service=travis-ci -repotoken $COVERALLS_TOKEN" exited with 1.
```

**Comment by peterbourgon** [comment]
We tried this before: https://github.com/oklog/ulid/pull/10

@tsenart Has anything changed in your opinion, or does this PR bring new stuff to the table?

**Comment by bruth** [comment]
Sorry I missed! I guess my response to that discussion is that the package could provide a default method so it "just works", but obviously the user can choose if they want to go with it or wrap to serialize as a string instead.

**Comment by tsenart** [comment]
> @tsenart Has anything changed in your opinion, or does this PR bring new stuff to the table?

Yeah, my thoughts on this have changed.

> Sorry I missed! I guess my response to that discussion is that the package could provide a default method so it "just works", but obviously the user can choose if they want to go with it or wrap to serialize as a string instead.

I'm cool with that approach as long as those tradeoffs are clearly documented.

**Comment by tsenart** [comment]
So, for that matter, providing a default `Scan` implementation that also handles strings would be good. Then those who want to serialize strings only need to implement one method, not two, on their wrapper type.

**Comment by bruth** [comment]
Updated to handle scanning strings as well. Assuming this is good, I will quick add a couple tests.

File: ulid.go [hunk 1]
```diff
@@ -340,3 +341,24 @@ func (id *ULID) SetEntropy(e []byte) error {
...
+		return nil
+
+	case string:
+		return id.UnmarshalText([]byte(x))
+
+	case []byte:
+		return id.UnmarshalBinary(x)
+	}
+
+	return errors.New("ulid: source value must be bytes")
```
Comment (tsenart) [inline ulid.go]: With string support, this error message must be adapted.

File: ulid.go [hunk 2]
```diff
@@ -340,3 +341,24 @@ func (id *ULID) SetEntropy(e []byte) error {
 func (id ULID) Compare(other ULID) int {
 	return bytes.Compare(id[:], other[:])
 }
+
+// Scan implements the sql.Scanner interface.
+func (id *ULID) Scan(src interface{}) error {
+	switch x := src.(type) {
+	case nil:
+		return nil
+
```
Comment (tsenart) [inline ulid.go]: Please remove the vertical whitespace between the switch cases.

File: ulid.go [hunk 3]
```diff
@@ -340,3 +341,24 @@ func (id *ULID) SetEntropy(e []byte) error {
...
+		return id.UnmarshalText([]byte(x))
+
+	case []byte:
+		return id.UnmarshalBinary(x)
+	}
+
+	return errors.New("ulid: source value must be bytes")
+}
+
+// Value implements the sql/driver.Valuer interface.
```
Comment (tsenart) [inline ulid.go]: Let's add clear documentation in this comment that the generated value is of the []byte type and how to do it with strings via the wrapper type approach.

**Comment by tsenart** [comment]
Yep, +1 for tests.

**Comment by tsenart** [comment]
> Addressed the comments. For Value, I noted that a wrapper type can be created that calls MarshalText. I suppose a code example could be added or even a function that handles this automatically could do it, something like:

I'd just provide a code example in GoDoc.

File: ulid_test.go [hunk 4]
```diff
@@ -390,6 +390,32 @@ func TestOverflowHandling(t *testing.T) {
 	}
 }
 
+func TestScanBytes(t *testing.T) {
```
Comment (tsenart) [inline ulid_test.go]: These two tests are mostly identical. Could we merge them into a single table driven test? Something like:

```go
id := ulid.MustNew(123, crand.Reader)

for _, tc := range []struct{
    name string
    in interface{}
    out ULID
    err error
} {
    {"string", id.String(), id, nil},
    {"bytes", id[:], id, nil},
    {"nil", nil, ULID(nil), nil},
    {"other", 44, nil, ErrScanValue},
} {
    tc := tc
    t.Run(tc.name, func(t *testing.T) {
        t.Parallel()

        var out ulid.ULID
        err := out.Scan(tc.in)
        if got, want := out, tc.out; !got.Equal(want) {
            t.Errorf("got ULID %s, want %s", got, want)
        }

         if got, want := fmt.Sprint(err), fmt.Sprint(tc.err); got != want {
             t.Errorf("got err %q, want %q", got, want)
         }
    })
    
}
```

**Comment by bruth** [comment]
Good call with the tests. Updated. I also added the example in GoDoc. This is what is looks like:

screen shot 2018-07-12 at 7 26 25 am

**Comment by tsenart** [comment]
Ran tests locally. They pass. I'll fix CI today.

**Comment by imaramos** [comment]
Just a question, I am storing this in postgres as bytea, I have tried to do encoding in order to be more readable to a human but I am not being able to do it. Shouldn't at least to be able to see the numeric value that the byte array represents?
Thank you

**Comment by imaramos** [comment]
I am storing the ulid byte array , as a bytea column in postgres. So I have the byte array stored. But when doing the query it isnt human readable ( because it is only bytes). I am asking if there is a way of show at db level this column as a more human readable value, allowing also do the match with the ULID representation. ( I think I will not be able to do that because  the encoding used is Crockford's base32)

**Comment by tsenart** [comment]
If you store bytes in Postgres, you’ll get bytes back when querying. If you need the database to show the string representation of the ULIDs, you must either store that instead or define a Postgres function that can do the encoding itself at query time. For the letter, I found this on Google: https://github.com/schinckel/ulid-postgres

**Comment by tsenart** [comment]
I’d recommend you simply store the string representation unless you have space constraints true justify the gains of storing bytes.

**Comment by imaramos** [comment]
I have to try that , but it is still in progress so...  About the storing the string, is not about the space but also the efficience and also I will loose the ability to sort right?

**Comment by peterbourgon** [comment]
ULID strings are sortable. Whether or not it's efficient enough for your application is a question that can only be answered with representative usage data and profiling.

**Comment by imaramos** [comment]
Right, thank you for your help guys!

**Pull request merged**